### DISCUSSION TITLE
How to avoid averaging at the boundary of two materials.

### DISCUSSION THREAD

**Discussion by vlamor** [NanoComp/meep#2832] [Q&A]
Using MEEP I simulate the passage of light through a bent optical fiber. Optical fiber (2D) consists of a cladding and a core . I created the cladding and core using the gdstk library. An example geometry is shown in Fig. 1.
![time_500_after_50_res_10_ang_30 0__decay_0 1_nfreq_50_delta_0 1](https://github.com/NanoComp/meep/assets/13791658/6dff9e34-925d-4f1c-a321-754becb1ef11)
 Due to averaging at the boundary of two materials, there is no sharp jump in the refractive index, as is the case in reality. And therefore, total internal reflection in the core does not occur (as it shown in above Figure). To avoid averaging, I did the following steps:
1. Set eps_averaging = False in the simulation.
2. Created geometry in two ways:
      a) by simply applying the core to the shell: cladding + core
      b) from the cladding produced a Boolean subtraction of the core. Increased the radius of the core by various values (from 0 to 10E-8).
3. I wrote epsilon to an .h5 file. Changed the average values at the boundary to the values of the refractive index of the shell. And read this data back into the simulation. Averaging occurred again.
4. Added a dft field with the parameter yee_grid = true. 

None of these options worked: the refractive index is averaged at the core-cladding boundary, and because of this averaging, total internal reflection does not occur even for small bending angles, which should happen in reality. Is exists an effective way to avoid averaging at the boundary of two materials in MEEP?

**Comment by oskooi** [comment]
This could be due to either (1) a real physical effect or (2) discretization artifacts from the finite-grid resolution.

Regarding (1), what is the band structure for this waveguide? Is it single or multimode at this wavelength? For a single-mode waveguide, you can try making a plot of the bending loss ($1 - R - T$) vs. radius of curvature of the bend. 

Regarding (2), what happens to the results as you repeatedly *double* the grid resolution? See [FAQ/Checking convergence](https://meep.readthedocs.io/en/latest/FAQ/#checking-convergence).

Also, have you confirmed that your mode source has been set up correctly? As a simple test, a *straight* waveguide (no bend) should produce a transmittance of one for the guided mode. For reference, see [Tutorial/Index-Guided Modes in a Ridge Waveguide](https://meep.readthedocs.io/en/latest/Python_Tutorials/Eigenmode_Source/#index-guided-modes-in-a-ridge-waveguide).

**Reply by vlamor** [reply]
More details. This is multimode optical fiber. refractive indices of core is 1.49893141723397, cladding is 1.45411616508421. The core diameter is 50 um, the cladding diameter is 125 um, total length of the fiber is 250 um. Resolution is 50 pixels/um. I am making bending by gdstk command (where upper points of fiber is rotated 30 degrees relative to the center point of the fiber)

gdstk.FlexPath(points, width=125, bent_radius=100, layer=1, simple_path=True) 

So, bent radius is equal to 100. Wavelength is equal to 1 um in my simulation. And I checked that 1 um in MEEP corresponds to 1 unit of length in gdstk library. 

Reflection of the straight fiber is approximately 3%. Transmittance of the light which trapped by fiber is close to 100%.

Regarding (2). Application running in parallel regime. A maximum resolution is 50 pixels/um. More value of the resolution leads to lack of memory.

My source is as follows:
```
src_vol = mp.GDSII_vol(gdsII_file, SOURCE, si_zmin, si_zmax)

# Focused source
rot_angle = 0  # CCW rotation angle about z axis (0: +y axis)
beam_kdir = mp.Vector3(0, 1, 0).rotate(
    mp.Vector3(0, 0, 1), math.radians(rot_angle))

sources = [mp.GaussianBeamSource(
                        src=mp.ContinuousSource(frequency=frequency)
                        volume=src_vol, 
                        beam_x0 = mp.Vector3(0, 20.0),
                        beam_kdir=beam_kdir,
                        beam_w0=20, # beam waist radius
                        beam_E0 = mp.Vector3(0, 0, 1),
                        )
          ]
```
I have to simulate wavelengths in range 750 - 1100 um.

**Comment by stevengj** [comment]
(It converges to a sharp jump in the limit of infinite resolution.  The averaging actually can help it converge to the discontinuous result *faster*.)

**Comment by stevengj** [comment]
If this is modelling a silica-like optical fiber, with a ≈1% index contrast, then your bend is likely to be too sharp to prevent light from escaping — it looks like your bending radius is only a few wavelengths.

**Reply by vlamor** [reply]
Yes, this is a silica-like optical fiber: refractive indices of core is 1.49893141723397, cladding is 1.45411616508421.  The core diameter is 50 um, the cladding diameter is 125 um, total length of the fiber is 250 um. Resolution is 50 pixels/um. I am making bending by  gdstk command (where upper points of fiber is rotated 30 degrees relative to the center point of the fiber)

`gdstk.FlexPath(points, width=125, bent_radius=100, layer=1, simple_path=True) `

So, bent radius is equal to 100. Wavelength is equal to 1 um in my simulation. And I checked that 1 um in MEEP corresponds to 1 unit of length in gdstk library. Please correct me if I am wrong.

**Reply by stevengj** [reply]
If you zoom in on the image you plotted, it looks like the waveguide is straight + short bend region + straight, where the short bend region is only a few wavelengths long:

![image](https://github.com/NanoComp/meep/assets/2913679/a7dffdf6-93cb-4705-80ef-70abdd85d5a5)

**Reply by vlamor** [reply]
Thank you for the quick reply. I will try larger bent radius. And may be you have an idea how to construct more smooth transition?

**Comment by stevengj** [comment]
(It's also not clear if you are launching a fiber mode, or just a Gaussian beam or something?)
